### PR TITLE
Alpine support

### PR DESCRIPTION
**Pull request by chonton**
### Type of change

_Select the type of your PR_

- Enhancement / new feature
- Refactoring

### Description

Smaller image sizes with Alpine base

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

**Comment by chonton** [comment]
Comparison of image sizes:
- centos is current image
- % column is the percentage of the image size compared to centos 

|Image<br/>(Size in MB)|centos<br/>Java8|debian<br/>Java8|%|debian<br/>Java11|%|Alpine<br/>Java8|%|Alpine<br/>Java11|%|
|:----|----:|----:|----:|----:|----:|----:|----:|----:|----:|
|cluster-operator|384|273|71%|152|40%|213|55%|91.4|24%|
|entity-operator-stunnel|226|101|45%|101|45%|8.4|4%|8.4|4%|
|java-base|362|251|69%|130|36%|192|53%|69.7|19%|
|kafka|426|310|73%|189|44%|250|59%|129|30%|
|kafka-base|426|310|73%|189|44%|250|59%|129|30%|
|kafka-connect|426|310|73%|189|44%|250|59%|129|30%|
|kafka-connect-s2i|426|310|73%|189|44%|250|59%|129|30%|
|kafka-init|377|266|71%|145|38%|206|55%|84.5|22%|
|kafka-mirror-maker|426|310|73%|189|44%|250|59%|129|30%|
|kafka-stunnel|226|101|45%|101|45%|8.4|4%|8.4|4%|
|stunnel-base|226|101|45%|101|45%|8.4|4%|8.4|4%|
|test-client|426|310|73%|189|44%|250|59%|129|30%|
|topic-operator|388|277|71%|156|40%|217|56%|95.4|25%|
|user-operator|401|290|72%|169|42%|230|57%|108|27%|
|zookeeper|426|310|73%|189|44%|250|59%|129|30%|
|zookeeper-stunnel|226|101|45%|101|45%|8.4|4%|8.4|4%

**Comment by scholzj** [comment]
Could you please clarify what does this PR exactly do? It seems to me on a first look that it is removing CentOS. But we do want to have CentOS as a base image.

**Comment by chonton** [comment]
The intent is to make images smaller.  There are multiple approaches that can be used.  My steps were:
1. Change to a Debian minimal base image
a. Java8 - package just jre (25 to 55% savings)
b. Java11 - use jlinker to package just required components (55 to 60% savings)
2. Change to Alpine base image
a. Java8 - package just jre (41 to 96% savings)
b. Java11 - use jlinker to package just required components (70 to 96% savings)

What features of Centos makes it desirable for a base image?

**Comment by scholzj** [comment]
Thanks for the PR. But using CentOS makes it easy for us to run it on RHEL. Neither Debian nor Alpine will do that. More over using multiple OSes will significantly increase the maintenance effort.

**Comment by chonton** [comment]
I don't follow you.  How does the container OS make it easier or harder to run on a particular Host OS?  For example, I am developing on a Mac with the Docker Daemon running on a VM.  The VM is running a rather generic 4.15.0 kernel.

See [Docker "Container Host" vs. "Container OS"](http://www.floydhilton.com/docker/2017/03/31/Docker-ContainerHost-vs-ContainerOS-Linux-Windows.html) and [Docker Base Image OS Size Comparison](https://www.brianchristner.io/docker-image-base-os-size-comparison/)

**Comment by scholzj** [comment]
I'm not talking about RHEL as the host-OS (which is more or less unrelated as you said) but about RHEL as the Container base OS.

**Comment by chonton** [comment]
OK, makes sense to support a limited number of base OS. (preferably one) . What are the selection criteria for choosing the container OS?  For me, the criteria are:
1. stable functionality
2. boot time of kubernetes pod

**Comment by scholzj** [comment]
Basically, we work for Red Hat and use this project within Red Hat where we obviously use RHEL. So if we move this project to other OS, we will still need to do all the work to make it run on RHEL. So that is the reason why we definitely want to have CentOS (as the more or less unsupported version of RHEL) here.

And to be honest, I do not remember too many complains from the users for example on Slack about us using CentOS or about the image size. Kafka it self is fairly big and heavy so maybe most users prioritise other aspects.

After your PR, we had some internal discussion about supporting Alpine as an alternative. But to be honest, I do not think we have currently the bandwidth to do that. CentOS and Alpine are very different OS and developing, maintaining and testing on both would in our opinion cost too much effort.

If you really want to build your own images using something like Alpine Linux, we would like to make it as easy as possible for you - for example by modyfying the build to allow you to easily swap the Docker files or something like that. So if you have any ideas about that, feel free to share them with us.

**Comment by tombentley** [comment]
> > maintaining and testing on both would in our opinion cost too much effort
>
> I'm not convinced that the matching the container OS with host OS decreases the testing load.

It's not really about the host OS, it's about the opportunity cost...

We use RHEL as the container OS because all Red Hat cloud native apps are based on RHEL (why? Basically economies of scale). If the Strimzi project were offering two flavours of images the project would need the resources to test both flavours. Alternatively if the Strimzi project completely switched to basing its images on something very different from RHEL we (as in the Red Hatters working on Strimzi) would need to spend extra time making sure it worked on RHEL (i.e. we'd have to maintain our own dockerfiles and do even more testing). In either case there is more work for Red Hat.

However, the reality is that most of the people contributing to Strimzi are paid by Red Hat. So more work for Red Hat in productizing Strimzi means less time actually working on Strimzi (e.g. slower rate of new features). This would be detrimental the Strimzi project, even if it might be beneficial (in the short term) to some of our users (such as you). In the longer term those same users would likely experience a Strimzi that had fewer features and more bugs.

File: .travis.yml [hunk 1]
```diff
@@ -36,3 +36,6 @@ env:
...
   - secure: pVE+P4F7b14BhGBUXYhPiTBYwsWj50jOQ76Lwae5IcWFKB8Ibrr8bvIW/bLfnsjVv51d0Bf0vyGRHoW2V2POC6n1wkzE3nHZ+lSJQkmg5jsqr9z4Nxd6EsZUpwlSBC7/hzlqbE8GeBeIg5qxoclInZrcG+zoTgcfA1N/hIooNniP3TE1jBkRueYptvAiG8ugL9IwbQ8UvXXE9GVHkROwax4AOEFG7ZjWDzOH2pzvGtvT34pPDGtL0ZYWwbklZsdkCkFmHR0Lo8vQi1sesqi3KmiqStlrRfEHw+RgRhk2tgB76AzIyUO4zOX0+nGqKWwfTGt933/19ENrg+uKJr1j57N9Cwzh2xxRGlwUi44c64lmOCMtWedU+xAF8omVvoOuqYibDTeNPm8lYaiV0JWoxrJ9vKwh3jEyi9lbUmYpU7I+Pi5SC2crQQQDYH/Qfd+gKY/V7QfCKfjjneKe/TA0uXBrHTv07E+WLGcLFVQ6ir3cvzk1lU5iyH2cR1mfCMY5/DN7EQQOJn5Ww1dwyJBPvJkBGHV/hR3QZUWtcE7nuvmbrFMZQBLzlnYuZfMtdRuyU8dOUtsDrt0On1A1JswvOfwAqSyitQ7mAUlq2mqazMRTNFmj8zVPZ12IWuMsEw4zTc16RAedszME166B2fp9PEMF4oErGlHxoBqhrpanyO0=
+  matrix:
```
Comment (chonton) [inline .travis.yml]: This change probably not wanted in master branch.

File: docker-images/kafka-base/Dockerfile [hunk 2]
```diff
@@ -1,41 +1,38 @@
-FROM centos:7
-
-ARG JAVA_VERSION
+FROM strimzi/java-base:latest
```
Comment (tombentley) [inline docker-images/kafka-base/Dockerfile]: Open question: What are the pros and cons of basing the `kafka-base` on the `java-base`?

File: .travis.yml (hunk 1 above)
Comment (tombentley) [reply .travis.yml]: Yeah, I agree TBH.

File: docker-images/kafka-base/Dockerfile [hunk 3]
```diff
@@ -1,41 +1,38 @@
...
+RUN yum -y install gettext nmap-ncat && yum clean all -y
 
 # set Kafka home folder
 ENV KAFKA_HOME=/opt/kafka
 
-# Add kafka user with UID 1001
+# Add kafka user with UID 1002
 # The user is in the group 0 to have access to the mounted volumes and storage
-RUN useradd -r -m -u 1001 -g 0 kafka
+RUN useradd -r -m -u 1002 -g 0 kafka
```
Comment (scholzj) [inline docker-images/kafka-base/Dockerfile]: I think this change could cause problems with access rights to storage etc. Please keep the user as `1001`.

**Comment by scholzj** [comment]
@chonton 

> I understand your point about deploying on RHEL. I'm not convinced that the matching the container OS with host OS decreases the testing load.

I do not talk about host OS at all. We do not do any special tests on different host OS types. We expect the containers to do their _magic_. When I talk about RHEL I'm specifically referring to RHEL container base image - not host OS.

The increased maintenance, development and testing effort comes from the fact that we would have two separate sets of images which need to be maintained and tested. From my point of view, if it is in our repository, we should be taking care of it. Which I'm afraid doesn't make me entirely happy with just moving the alpine Docker files into a separate subdirectory. If they are part of the repo then we should be taking care of them and I do not think we have the resources to do that.

**Comment by chonton** [comment]
Splitting this into multiple PRs to simplify discussions.
#1294, #1295, #1296

**Pull request closed without merging**